lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
     ch = ContextHandler = {
       contextPreviewExists: false
       UPDATE_DELAY: 1000
+      LABEL_SEQUENCE: [
+        {type: "keyword", value: /^\\label$/}
+        {type: "lparen", value: /^\{$/}
+        {type: "variable.parameter", value: /.*/}
+        {type: "rparen", value: /^\}$/}
+      ]
+      LABEL_PARAMETER_INDEX: 2
 
       getWholeEquation: (range) ->
-        editor.getSession().getTextRange(range)
+
+        tokenValues = []
+        labelSequenceIndex = 0
+        labelParameters = []
+        curLabelParameter = null
+        curLabelTokens = []
+
+        tokenIterator = new TokenIterator(editor.getSession(), range.start.row, range.start.column)
+        tokenIterator.stepForward() # at first it should be on the closing token of begin sequence
+        {row: curRow, column: curColumn} = tokenIterator.getCurrentTokenPosition()
+        token = tokenIterator.getCurrentToken()
+
+        while range.contains(curRow, curColumn + token.value.length)
+          tokenToMatch = ch.LABEL_SEQUENCE[labelSequenceIndex]
+
+          if token.type == tokenToMatch.type and tokenToMatch.value.test(token.value)
+
+            if labelSequenceIndex == ch.LABEL_PARAMETER_INDEX
```

Comment:
What do you think of more simple and probably more useful function which searches for a given keyword and removes that keyword and everything inside a properly balanced pair of curly brackets afterwards? I mean, given text `"\foo\bar\label{\foo{}\bar{}}\baz\qux"`, iterator positioned at the first token, and keyword `\label` it searches for `token.type=keyword` and `value=\label`, then runs `findSurroundingBrackets` from the position just after `\label{` and removes both `\label` and the range returned by `findSurroundingBrackets`

We coud reuse such code in other similar scenarios.